### ISSUE TITLE
Replace regex-based code categorization with CoreML MLTextClassifier (1.4.0)

### ISSUE THREAD

**Issue by MagnetonIO** [MagnetonIO/copied-app#54]
## Why

After 1.3.x we have a stack of regex/heuristic patches:

* MagnetonIO/copied-app#49 (markdown → swift on "Swift" mention)
* MagnetonIO/copied-app#53 (Auto-update → SQL, Defer → Zig, README → Lua, Python → TOML, YAML → Shell)
* 1.3.2 patches: strong/weak anchor split, NLTagger prose disqualifier, code-language disqualifier in config detector, reordered chain, tightened SQL/Zig/Ruby/Python/TS/Java/C/Lua/C++ regexes

Each fix surfaces a new edge case. The root cause is that regex pattern matching has no notion of distributional context — a single word like "lambda" or "update" or "defer" can't be disambiguated by neighboring tokens via regex alone. Bag-of-keywords classifiers (`detectLanguageByKeywords`) have the same blind spot.

ML-based text classification handles this natively: the model learns that "lambda" surrounded by `:`, parentheses, and arithmetic operators means Python; surrounded by Greek letters and "calculus" means prose. We can't teach that to a regex.

This ticket replaces the entire heuristic stack with one CoreML classifier.

## Approach

**Apple's CreateML** `MLTextClassifier` trains a small (typically 0.5–2 MB) on-device text classifier from labeled examples. Inference runs in \~5 ms per snippet, fully on-device, no network. Apple uses MaxEnt over n-gram features by default; can opt into transformer if accuracy demands.

The model replaces \~600 lines of `CodeDetector` regex/heuristic code with one `prediction(text:)` call.

### Architecture (after migration)

```
Tier 0: UTI source-code        — kept (definitive when present)
Tier 1: Indisputable headers   — kept (~5 patterns: `<?php`, `<!DOCTYPE html>`,
                                  shebangs, `@import("std")`)
Tier 2: CoreML classifier       — NEW: returns (label, confidence)
Tier 3: Confidence < 0.6 → text — fallback
```

That's it. No anchor regex chain, no keyword bags, no NLTagger disqualifier, no markdown scorer, no config-format detector. All of it collapses into the model.

### Class taxonomy (\~30 labels)

**Programming languages (24):** swift, python, javascript, typescript, java, kotlin, c, cpp, csharp, go, rust, ruby, php, scala, perl, lua, haskell, elixir, erlang, clojure, dart, r, sql, shell

**Markup / config (6):** markdown, html, css, xml, json, yaml, toml, dockerfile, makefile

**Other (2):** plain_text, prose

Total: \~32 classes. Within `MLTextClassifier`'s sweet spot.

Optional Phase 2: emerging-language labels (zig, mojo, julia, ocaml, v, nim, gleam, fsharp). Add when training corpus available.

### Code samples

1. **GitHub Linguist** — [https://github.com/github-linguist/linguist](<https://github.com/github-linguist/linguist>)
   * `samples/<Language>/` directory has 100–1000 curated examples per language
   * MIT-licensed, \~50 MB total
   * **Best starting source** — Linguist already curated these for language detection
   * Extract: clone repo, walk `samples/`, treat directory name as label
2. **CodeSearchNet** — [https://github.com/github/CodeSearchNet](<https://github.com/github/CodeSearchNet>)
   * 6 languages × millions of functions: Python, Java, JavaScript, PHP, Ruby, Go
   * Apache 2.0, paired with natural-language docstrings
   * Use for realistic snippet-shaped data
   * Download: `aws s3 cp s3://code-search-net/CodeSearchNet/...`
3. **The Stack v2 (BigCode)** — [https://huggingface.co/datasets/bigcode/the-stack-v2](<https://huggingface.co/datasets/bigcode/the-stack-v2>)
   * 6 TB across 600+ languages, permissive licenses only (filtered)
   * Apache 2.0
   * Sample 1–2 K files per language to keep corpus manageable
   * Filter to files < 5 KB for snippet realism
4. **The Stack Smol** — [https://huggingface.co/datasets/bigcode/the-stack-smol](<https://huggingface.co/datasets/bigcode/the-stack-smol>)
   * Subset of The Stack, \~10 K files per language
   * Better starting point than full Stack

### Prose samples

5. **WikiText-103** — [https://huggingface.co/datasets/wikitext](<https://huggingface.co/datasets/wikitext>)
   * 100 M tokens of cleaned English Wikipedia
   * Apache 2.0
   * Chunk into 200–2000-char snippets for training
6. **OpenWebText** — [https://huggingface.co/datasets/Skylion007/openwebtext](<https://huggingface.co/datasets/Skylion007/openwebtext>)
   * 8 M web pages, MIT
   * Mix of articles, blogs, forum posts — closer to real clipboard content than Wikipedia alone

### Markdown samples

7. **GitHub READMEs** — scrape via GitHub API
   * Top 10 K repos by stars, fetch [README.md](<http://README.md>)
   * Already permissively licensed (open-source READMEs are public domain or covered by repo license)
8. **Awesome lists** — [https://github.com/sindresorhus/awesome](<https://github.com/sindresorhus/awesome>)
   * Curated lists in markdown, indexable from the awesome graph
   * Real-world markdown structure

### Config samples

 9. **GitHub Search API** — search for `.yaml`/`.toml`/`Dockerfile`/`Makefile` files
    * Sample top 1–2 K of each by star count
    * Permissive-license filter
10. **Existing Linguist samples** — `linguist/samples/{YAML,JSON,TOML,Dockerfile,Makefile}/`
    * Already curated, ready to use

### HTML/CSS/XML

11. **Common Crawl WET archives** — [https://commoncrawl.org/](<https://commoncrawl.org/>)
    * Petabyte-scale web corpus
    * WET = plain text (skip), WAT = metadata, WARC = full HTML
    * Sample 1 K HTML pages from various domains
    * Public-domain index

### Negative / boundary samples (the hard cases)

12. **Stack Overflow data dump** — [https://archive.org/details/stackexchange](<https://archive.org/details/stackexchange>)
    * Q&A pairs with mixed prose + code
    * CC BY-SA license
    * Critical for training the classifier on **real clipboard content** (questions paste both prose and code)
13. **Internal regression cases** — every false-positive we've reported:
    * "Auto-update for paid Stripe build…" → prose
    * "Out of scope - Tree-sitter / CoreML… Defer until…" → prose
    * "Lambda calculus: a model of computation" → prose
    * "Library systems offer end-of-day functions to Local users" → prose
    * "select wisely from the menu, then update your order" → prose
    * The full [README.md](<http://README.md>) from this repo → markdown
    * The Python `class PostgresUnifiedBackend` example → python
    * YAML files with embedded `run: |` shell → yaml

### Step 1 — Corpus ingestion (1 day)

```bash
mkdir -p corpus/{code,markdown,prose,config,html}

# Linguist (cheap, ~50 MB)
git clone https://github.com/github-linguist/linguist
rsync -av linguist/samples/ corpus/code/

# WikiText
huggingface-cli download wikitext --revision wikitext-103-v1 --local-dir corpus/prose/wikitext

# Stack Smol (per-language)
for lang in swift python javascript ...; do
  huggingface-cli download bigcode/the-stack-smol --include "data/$lang/*" --local-dir corpus/code/$lang
done

# READMEs (10 K)
gh api 'search/repositories?q=stars:>1000&per_page=100' | jq -r '.items[] | .full_name' \
  | head -10000 \
  | parallel 'gh api repos/{}/readme | jq -r .content | base64 -d > corpus/markdown/{=s|/|_|g=}.md'
```

### Step 2 — Snippet chunking + class balancing (1 day)

`scripts/build-categorization-corpus.swift`:

* Walk corpus tree, label by parent directory
* Chunk each file into 200–2000-char snippets (random window per file, max 5 chunks per file)
* Cap at 2000 examples per class (balance)
* Write JSONL: `{"text": "...", "label": "python"}`
* Output: `corpus/training.jsonl` (\~50 MB), `corpus/validation.jsonl` (\~10 MB)

### Step 3 — CreateML training (1 day)

CreateML.app or Swift playground:

```swift
import CreateML
import Foundation

let data = try MLDataTable(contentsOf: URL(fileURLWithPath: "corpus/training.jsonl"))
let (train, test) = data.randomSplit(by: 0.8)

let classifier = try MLTextClassifier(
    trainingData: train,
    textColumn: "text",
    labelColumn: "label",
    parameters: .init(algorithm: .maxEnt(.init(revision: 2)))
)

print("Training error: \(classifier.trainingMetrics.classificationError)")
print("Validation error: \(classifier.validationMetrics.classificationError)")

try classifier.write(to: URL(fileURLWithPath: "ClippingClassifier.mlmodel"),
                     metadata: .init(author: "Magneton Labs", version: "1.0"))
```

Tune `algorithm` between `.maxEnt`, `.dynamicEmbedding`, `.transferLearning(.dynamicEmbedding(...))` based on accuracy/size trade-off.

Acceptance criteria: ≥ 95% top-1 accuracy on held-out test set.

### Step 4 — Integration (1 day)

`CopiedKit/Sources/CopiedKit/Services/CodeDetector.swift` becomes:

```swift
import CoreML

public struct CodeDetector: Sendable {
    private static let model: ClippingClassifier = {
        let config = MLModelConfiguration()
        config.computeUnits = .cpuAndNeuralEngine
        return try! ClippingClassifier(configuration: config)
    }()

    public static func classify(_ text: String) -> (label: String, confidence: Double)? {
        guard text.count >= 5, text.count <= 64_000 else { return nil }

        // Tier 1: indisputable headers
        if let header = headerLanguage(text) { return (header, 1.0) }

        // Tier 2: CoreML
        guard let result = try? model.prediction(text: text) else { return nil }
        let label = result.label
        let confidence = result.labelProbabilities[label] ?? 0
        guard confidence >= 0.6 else { return nil }
        return (label, confidence)
    }

    private static func headerLanguage(_ text: String) -> String? {
        let head = text.prefix(200)
        if head.hasPrefix("<?php") { return "php" }
        if head.contains("<!DOCTYPE html") { return "html" }
        if head.hasPrefix("#!/bin/bash") || head.hasPrefix("#!/bin/sh") || head.hasPrefix("#!/bin/zsh") { return "shell" }
        if head.hasPrefix("#!/usr/bin/env python") { return "python" }
        if head.hasPrefix("#!/usr/bin/env node") { return "javascript" }
        if head.hasPrefix("#!/usr/bin/env ruby") { return "ruby" }
        return nil
    }

    // languageFromUTIs and codeEditorBundleIDs stay
    public static func languageFromUTIs(_ types: [String]) -> String? { /* unchanged */ }
    public static let codeEditorBundleIDs: Set<String> = [/* unchanged */]
    public static func defaultLanguageForIDE(_ bundleID: String) -> String? { /* unchanged */ }
}
```

`ClipboardService.categorize` becomes:

```swift
nonisolated static func categorize(...) -> (isCode: Bool, detectedLanguage: String?) {
    // 1. UTI definitive
    if let lang = CodeDetector.languageFromUTIs(types) { return (true, lang) }

    // 2. CoreML
    if let text, let result = CodeDetector.classify(text) {
        let isCode = !["plain_text", "prose"].contains(result.label)
        return (isCode, isCode ? result.label : nil)
    }

    // 3. IDE bundle hint
    if let bundleID = appBundleID, CodeDetector.codeEditorBundleIDs.contains(bundleID) {
        return (true, CodeDetector.defaultLanguageForIDE(bundleID))
    }

    return (false, nil)
}
```

### Step 5 — Test corpus + perf (1 day)

`Tests/CopiedKitTests/CodeDetectorTests.swift`:

* All MagnetonIO/copied-app#49 / MagnetonIO/copied-app#53 reported cases as regression checks
* 100+ real samples per class from a held-out test corpus
* Perf: `model.prediction(text:)` < 10 ms p99 in Release on M-series Mac and iPhone

### Step 6 — Ship 1.4.0 (1 day)

Bump MARKETING_VERSION 1.3.x → 1.4.0, build .pkg + TF, deploy. Bump V7 → V8 reclassify so existing rows re-tag through the new model.

## Critical files

* `corpus/` (NEW, gitignored — too large for git) — training data
* `scripts/build-categorization-corpus.swift` (NEW) — corpus chunking + balancing
* `CreateML/TrainClippingClassifier.swift` (NEW) — training script
* `CopiedKit/Resources/ClippingClassifier.mlmodel` (NEW) — bundled model (\~1 MB)
* `CopiedKit/Package.swift` — add `.process("Resources/ClippingClassifier.mlmodel")` to target resources
* `CopiedKit/Sources/CopiedKit/Services/CodeDetector.swift` — gut existing implementation, replace with `classify()` (\~80 lines down from \~600)
* `CopiedKit/Sources/CopiedKit/Services/ClipboardService.swift` — simplify `categorize`, bump V7 → V8

## What gets deprecated / deleted

* `CodeDetector.anchorLanguage` + 80+ regex patterns
* `CodeDetector.compiledStrongAnchors` / `compiledWeakAnchors` / `strongAnchorPatterns` / `weakAnchorPatterns`
* `CodeDetector.detectLanguageByKeywords` + per-language keyword bags
* `CodeDetector.detectConfigFormat` + YAML/TOML/Dockerfile/Makefile heuristics
* `CodeDetector.looksLikeMarkdown` + scoring
* `CodeDetector.htmlIsTrivialWrapper`
* `CodeDetector.looksLikeProseNL` + NLTagger usage (model handles prose detection natively)
* The strong/weak anchor split + ≥2-hits requirement

Net code reduction: \~600 lines deleted, \~80 lines added, +1 MB model resource.

## Verification

1. **Test corpus** (held-out, never seen during training): ≥ 95% top-1 accuracy
2. **Regression suite**: every MagnetonIO/copied-app#49 / MagnetonIO/copied-app#53 reported case passes
3. **Perf**: `scripts/perf-trace.sh capture-loop` — capture-roundtrip < 50 ms p99 (model inference is part of the budget)
4. **Bundle size**: < 2 MB model addition
5. **Behavior parity** for cases the regex got right: Swift / Python / SQL / Markdown / YAML samples from the corpus all classify correctly
6. **iOS**: same model ships in CopiedKit, same accuracy on A-series chips (might be marginally slower but still < 20 ms)

## Risks

* **Training data quality** dominates accuracy. Garbage-in problems are real — Linguist samples can be old or atypical. Mitigation: hand-review a 100-sample validation set per class before training.
* **Bundle size growth** if model > 2 MB. Mitigation: tune algorithm + vocabulary size; can also lazy-load model on first use (deferred \~50 ms cost).
* **Inference latency on older iPhones** could exceed budget. Mitigation: benchmark on iPhone 12 / SE; fall back to header-only classification if model load fails.
* **Languages we don't train** become unrecognized. Mitigation: ship with "unknown_code" class; add languages quarterly.
* **Reproducibility**: training must be deterministic. Document corpus version, hyperparameters, CreateML version, output checksum. Store training script + corpus manifest in repo.

## Out of scope

* **SwiftTreeSitter for syntactic verification** — Phase 2. Useful for top 5 languages where we want sub-1% error.
* **Active learning loop** — capture mis-classified clippings (with user opt-in) for next training round. Phase 3.
* **Per-user fine-tuning** — overkill for a clipboard manager.
* **Online model updates** — over-the-air model refresh. Defer; quarterly app-update cadence is fine.

## Estimate

* Day 1: Corpus ingestion (Linguist + Stack Smol + WikiText + READMEs)
* Day 2: Chunking pipeline + balancing + train/test split
* Day 3: CreateML training + hyperparameter tuning + eval
* Day 4: Integration into CodeDetector / ClipboardService
* Day 5: Comprehensive test corpus, regression suite, perf bench
* Day 6: Ship 1.4.0 (Mac + iOS) + V8 migration

**\~1 week of focused work.** Pays back on every future categorization regression we'd otherwise spend half a day on.

## Related

* MagnetonIO/copied-app#49, MagnetonIO/copied-app#53 — regex-era false positives this will eliminate
* 1.3.2 (commit e1acf87 + later) — final regex tightening before this rewrite
